### PR TITLE
SMP: fix crash when switch to new task which is still running

### PR DESCRIPTION
**Pull request by GUIDINGLI**
## Summary
```

cpu0 thread0:                        cpu1:
sched_yield()
nxsched_set_priority()
nxsched_running_setpriority()
nxsched_reprioritize_rtr()
nxsched_add_readytorun()
up_cpu_pause()
                                     IRQ enter
                                     arm64_pause_handler()
                                     enter_critical_section() begin
                                     up_cpu_paused() pick thread0
                                     arm64_restorestate() set thread0 tcb->xcp.regs to CURRENT_REGS
up_switch_context()
  thread0 -> thread1
arm64_syscall()
    case SYS_switch_context
     change thread0 tcb->xcp.regs
    restore_critical_section()
                                     enter_critical_section() done
                                     leave_critical_section()
                                     IRQ leave with restore CURRENT_REGS
                                     ERROR !!!

Reason:
As descript above, cpu0 swith task: thread0 -> thread1, and the syscall() execute slowly, this time cpu1 pick thread0 to run at up_cpu_paused(). Then cpu0 syscall execute, cpu1 IRQ leave error.

Resolve:
Move arm64_restorestate() after enter_critical_section() done

This is a continued fix with:
https://github.com/apache/nuttx/pull/6833
```

## Impact

SMP

## Testing

qemu-armv8a:nsh_smp

## How to reproduce this issue:

add patch in:
https://github.com/apache/nuttx-apps/pull/2316

configure with:
./tools/configure.sh qemu-armv8a:nsh_smp -j8

run with qemu:
qemu-system-aarch64 -cpu cortex-a53 -smp 4 -nographic -machine virt,virtualization=on,gic-version=3 -net none -chardev stdio,id=con,mux=on -serial chardev:con -mon chardev=con,mode=readline -kernel ./nuttx

Console output:

```

ligd@Opt:~/platform/mainline/nuttx$ qemu-system-aarch64 -cpu cortex-a53 -smp 4 -nographic -machine virt,virtualization=on,gic-version=3 -net none -chardev stdio,id=con,mux=on -serial chardev:con -mon chardev=con,mode=readline -kernel ./nuttx
- Ready to Boot Primary CPU
- Boot from EL2
- Boot from EL1
- Boot to C runtime for OS Initialize
- Ready to Boot Second CPU
- Boot from EL2
- Boot from EL1
- Boot to C runtime for OS Initialize

NuttShell (NSH) NuttX-10.4.0
nsh> 
nsh> 
nsh> 
nsh> ps
  PID GROUP CPU PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK           STACK   USED  FILLED COMMAND
    0     0   0   0 FIFO     Kthread   - Assigned           0000000000000000 016368 001248   7.6%  CPU0 IDLE
    1     1   1   0 FIFO     Kthread   - Running            0000000000000000 016368 000448   2.7%  CPU1 IDLE
    2     2 --- 192 RR       Kthread   - Waiting  Semaphore 0000000000000000 016304 001184   7.2%  hpwork 0x403c9bd8 0x403c9c00
    3     3   0 100 RR       Task      - Running            0000000000000000 016336 003216  19.6%  nsh_main
    4     3 --- 100 RR       pthread   - Ready              0000000000000000 016368 001392   8.5%  wdt_kick0 0x402933e0 0
    5     3 --- 100 RR       pthread   - Waiting  Signal    0000000000000000 016368 001392   8.5%  wdt_kick1 0x402933e0 0
nsh> osperf
OS performance args: count:100, detail:false
==============================================================
Describe                                Max        Min        Avg
pthread-create                      1000000          0     100000
pthread-switch                      1000000          0      20000
[CPU1] _assert: Current Version: NuttX  10.4.0 29e7f6fadd Feb 27 2024 23:31:46 arm64
[CPU1] _assert: Assertion failed panic: at file: common/arm64_fatal.c:375 task(CPU1): osperf process: osperf 0x4027fe1400000000
[CPU1] up_dump_register: stack = 0x4027fe34
[CPU1] up_dump_register: x0:   0x403f73f0          x1:   0x403d7e80
[CPU1] up_dump_register: x2:   0x403c8518          x3:   0x403c95c0
[CPU1] up_dump_register: x4:   0x402ae204          x5:   0x0
[CPU1] up_dump_register: x6:   0x0                 x7:   0x0
[CPU1] up_dump_register: x8:   0x0                 x9:   0x0
[CPU1] up_dump_register: x10:  0x0                 x11:  0x0
[CPU1] up_dump_register: x12:  0x0                 x13:  0x16
[CPU1] up_dump_register: x14:  0x0                 x15:  0x9000000
[CPU1] up_dump_register: x16:  0x0                 x17:  0x0
[CPU1] up_dump_register: x18:  0x0                 x19:  0x403c8670
[CPU1] up_dump_register: x20:  0x0                 x21:  0x0
[CPU1] up_dump_register: x22:  0x0                 x23:  0x40280090
[CPU1] up_dump_register: x24:  0x403d8000          x25:  0x40281760
[CPU1] up_dump_register: x26:  0x0                 x27:  0x0
[CPU1] up_dump_register: x28:  0x0                 x29:  0x403d7fb0


```

### CODE REVIEW

**Comment by GUIDINGLI** [comment]
@masayuki2009 
Please review this

**Comment by masayuki2009** [comment]
> @masayuki2009 Please review this

@GUIDINGLI 
I will try this PR later.

**Comment by masayuki2009** [comment]
@GUIDINGLI 

```
common/riscv_cpupause.c: In function 'up_cpu_paused_restore':
common/riscv_cpupause.c:208:3: warning: implicit declaration of function 'arm_restorestate'; did you mean 'riscv_restorestate'? [-Wimplicit-function-declaration]
  208 |   arm_restorestate(tcb->xcp.regs);
      |   ^~~~~~~~~~~~~~~~
      |   riscv_restorestate
```

**Comment by masayuki2009** [comment]
@GUIDINGLI 

The following error happened with rv-virt:smp64 and rv-virt:smp.

```
nsh> ostest
...
fpu_test: Started task FPU#2 at PID=12
FPU#2: pass 1
FPU#1: pass 2
ERROR FPU#2: save1 and save2 do not match
Values before waiting (save1) (0x80029804):
    0000: 800149fe 800149fe 80031790 ffffffff 00000000 00000023 00000009 00000009
    0008: 80029990 00000002 80029804 fffffffe 00000000 80029b98 00000001 80029804
    0010: 00000001 00000009 80029804 00000001 00000001 800294cc 00000000 00000000
    0018: 00000000 00000000 00000000 00000000 0000002a 0000002d 00000030 00000020
    0020: 80006088 00000000 00000000 00000000 00000000 00000000 00000000 00000000
    0028: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
    0030: 00000000 407a0749 ffffffff 2336023d 400f40e9 00000000 00000000 00000000
    0038: 00000000 407a0749 ffffffff 403a0749 ffffffff 40fcd6b6 ffffffff 40c90fd0
    0040: ffffffff 00000000 00000000 00000000 00000000 40490fd0 ffffffff 3fcf1b9b
    0048: ffffffff 402df84d ffffffff 3f800000 ffffffff f01b866e 400921f9 6cdf266c
    0050: 3ff9e373 95aaf790 4005bf09 00000000 3ff00000 00000000 00000000 00000000
    0058: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
    0060: 00000000 00000001 ffffffff

Values after waiting (save2) (0x80029990):
    0000: 80014a76 80014a76 80031790 ffffffff 00000000 00000001 00000009 00000009
    0008: 80029990 00000002 80029990 fffffffd 00000000 80029b98 00000002 00000000
    0010: 00000000 00000009 80029804 00000001 00000001 800294cc 00000000 00000000
    0018: 00000000 00000000 00000000 00000000 0000002a 0000002d 00000030 00000020
    0020: 80006088 deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef
    0028: deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef
    0030: deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef
    0038: deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef
    0040: deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef
    0048: deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef
    0050: deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef
    0058: deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef 00000000
    0060: 00000000 00000004 ffffffff
```

**Comment by masayuki2009** [comment]
>The following error happened with rv-virt:smp64 and rv-virt:smp.

@GUIDINGLI 

The following crash happened with maix-bit:smp on dev board.

```
user_main: /dev/null test
dev_null: Read 0 bytes from /dev/null
dev_null: Wrote 1024 bytes to /dev/null

End of tes[CPU1] riscv_exception: EXCEPTION: Illegal instruction. MCAUSE: 0000000000000002, EPC: 0000000080019dc8, MTVAL: 00000017e03344d4
[CPU1] riscv_exception: PANIC!!! Exception = 0000000000000002
[CPU1] _assert: Current Version: NuttX  12.0.0 c75359849d Feb 28 2024 15:00:52 risc-v
[CPU1] _assert: Assertion failed panic: at file: :0 task(CPU1): FPU#1 process: FPU#1 0x80019d6c
[CPU1] up_dump_register: EPC: 0000000080019dc8
[CPU1] up_dump_register: A0: 0000000000000000 A1: fffffffffffffffd A2: 0000000000000000 A3: 0000000080403d88
[CPU1] up_dump_register: A4: 0000000000000002 A5: 0000000000000440 A6: 0000000000000000 A7: 0000000000000000
[CPU1] up_dump_register: T0: 0000000000000000 T1: 0000000000000000 T2: 0000000000000000 T3: 0000000000000000
[CPU1] up_dump_register: T4: 0000000000000000 T5: 0000000000000000 T6: 0000000000000000
[CPU1] up_dump_register: S0: 0000000000000000 S1: 0000000000000001 S2: 0000000000000000 S3: 0000000000000000
[CPU1] up_dump_register: S4: 0000000000000000 S5: 0000000080403480 S6: 0000000000000000 S7: 0000000000000000
[CPU1] up_dump_register: S8: 0000000000000000 S9: 0000000000000000 S10: 0000000000000000 S11: 0000000000000000
[CPU1] up_dump_register: SP: 000000008040c800 FP: 0000000000000000 TP: 0000000000000000 RA: 0000000080019dba
[CPU1] dump_stack: User Stack:
[CPU1] dump_stack:   base: 0x8040c0f0
[CPU1] dump_stack:   size: 00002000
[CPU1] dump_stack:     sp: 0x8040c800
[CPU1] stack_dump: 0x8040c7e0: 00000000 00000000 00000001 00000000 00000000 00000000 80019dba 00000000
[CPU1] stack_dump: 0x8040c800: deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef
[CPU1] stack_dump: 0x8040c820: deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef
[CPU1] stack_dump: 0x8040c840: deadbeef deadbeef deadbeef deadbeef 00000000 00000000 00000000 00000000
[CPU1] stack_dump: 0x8040c860: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[CPU1] stack_dump: 0x8040c880: 80019d6c 00000000 800064f8 00000000 8040c0d0 00000000 00000001 00000000
[CPU1] stack_dump: 0x8040c8a0: 00000000 00000000 8000373e 00000000 00000000 00000000 00000000 00000000
[CPU1] stack_dump: 0x8040c8c0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[CPU1] dump_tasks:    PID GROUP   CPU PRI POLICY   TYPE    NPX STATE   EVENT      SIGMASK          STACKBASE  STACKSIZE      USED   FILLED    COMMAND
[CPU1] dump_tasks:   ----   ---  ---- --- -------- ------- --- ------- ---------- ---------------- 0x80400260      2048      1128    55.0%    irq
[CPU1] dump_task:       0     0     0   0 FIFO     Kthread - Running            0000000000000000 0x80403eb0      2032      1120    55.1%    CPU0 IDLE
[CPU1] dump_task:       1     1     1   0 FIFO     Kthread - Assigned           0000000000000000 0x80401270      2032      1088    53.5%    CPU1 IDLE
[CPU1] dump_task:       2     2     0 100 RR       Task    - Waiting Semaphore  0000000000000000 0x804066f0      3024      2176    71.9%    nsh_main
[CPU1] dump_task:      39    39     0 100 RR       Task    - Waiting Semaphore  0000000000000000 0x8040aae0      2000      1456    72.8%    ostest
[CPU1] dump_task:      40    40     0 100 RR       Task    - Waiting Semaphore  0000000000000000 0x80407a30      8096      1520    18.7%    ostest Arg1 Arg2 Arg3 Arg4
[CPU1] dump_task:      41    41     1 100 RR       Task    - Running            0000000000000000 0x8040c0f0      2000       720    36.0%    FPU#1
^
```

**Comment by GUIDINGLI** [comment]
@masayuki2009 

Thanks for the feeding back.

I have updated the PR, please try again.

File: arch/arm/src/armv7-a/arm_cpupause.c [hunk 1]
```diff
@@ -168,10 +212,8 @@ int up_cpu_paused(int cpu)
    */
 
   arm_restorestate(tcb->xcp.regs);
-  spin_unlock(&g_cpu_wait[cpu]);
-  spin_unlock(&g_cpu_resumed[cpu]);
 
-  return OK;
+  return 0;
```
Comment (acassis) [inline arch/arm/src/armv7-a/arm_cpupause.c:216]: Please use OK instead 0. As you wrote in the Retuned Value: "On success, OK is returned." :-)

File: arch/arm/src/armv7-r/arm_cpupause.c [hunk 2]
```diff
@@ -131,6 +123,38 @@ int up_cpu_paused(int cpu)
 
   arm_savestate(tcb->xcp.regs);
 
+  return 0;
```
Comment (acassis) [inline arch/arm/src/armv7-r/arm_cpupause.c:126]: Ditto

File: arch/arm/src/armv7-r/arm_cpupause.c [hunk 3]
```diff
@@ -168,10 +212,8 @@ int up_cpu_paused(int cpu)
    */
 
   arm_restorestate(tcb->xcp.regs);
-  spin_unlock(&g_cpu_wait[cpu]);
-  spin_unlock(&g_cpu_resumed[cpu]);
 
-  return OK;
+  return 0;
```
Comment (acassis) [inline arch/arm/src/armv7-r/arm_cpupause.c:216]: Ditto

File: arch/arm/src/cxd56xx/cxd56_cpupause.c [hunk 4]
```diff
@@ -220,6 +205,45 @@ int up_cpu_paused(int cpu)
 
   arm_savestate(tcb->xcp.regs);
 
+  return 0;
```
Comment (acassis) [inline arch/arm/src/cxd56xx/cxd56_cpupause.c:208]: Ditto

File: arch/arm/src/cxd56xx/cxd56_cpupause.c [hunk 5]
```diff
@@ -251,10 +295,8 @@ int up_cpu_paused(int cpu)
    */
 
   arm_restorestate(tcb->xcp.regs);
-  spin_unlock(&g_cpu_wait[cpu]);
-  spin_unlock(&g_cpu_resumed[cpu]);
 
-  return OK;
+  return 0;
```
Comment (acassis) [inline arch/arm/src/cxd56xx/cxd56_cpupause.c:299]: Ditto

File: arch/arm/src/lc823450/lc823450_cpupause.c [hunk 6]
```diff
@@ -141,6 +133,38 @@ int up_cpu_paused(int cpu)
 
   arm_savestate(tcb->xcp.regs);
 
+  return 0;
```
Comment (acassis) [inline arch/arm/src/lc823450/lc823450_cpupause.c:136]: Ditto

File: arch/arm/src/lc823450/lc823450_cpupause.c [hunk 7]
```diff
@@ -173,11 +217,7 @@ int up_cpu_paused(int cpu)
 
   arm_restorestate(tcb->xcp.regs);
 
-  spin_unlock(&g_cpu_wait[cpu]);
-
-  spin_unlock(&g_cpu_resumed[cpu]);
-
-  return OK;
+  return 0;
```
Comment (acassis) [inline arch/arm/src/lc823450/lc823450_cpupause.c:220]: Ditto

File: arch/arm/src/rp2040/rp2040_cpupause.c [hunk 8]
```diff
@@ -181,6 +173,38 @@ int up_cpu_paused(int cpu)
 
   arm_savestate(tcb->xcp.regs);
 
+  return 0;
```
Comment (acassis) [inline arch/arm/src/rp2040/rp2040_cpupause.c:176]: Ditto

File: arch/arm/src/rp2040/rp2040_cpupause.c [hunk 9]
```diff
@@ -212,10 +256,8 @@ int up_cpu_paused(int cpu)
    */
 
   arm_restorestate(tcb->xcp.regs);
-  spin_unlock(&g_cpu_wait[cpu]);
-  spin_unlock(&g_cpu_resumed[cpu]);
 
-  return OK;
+  return 0;
```
Comment (acassis) [inline arch/arm/src/rp2040/rp2040_cpupause.c:260]: Ditto

File: arch/arm/src/sam34/sam4cm_cpupause.c [hunk 10]
```diff
@@ -174,10 +218,8 @@ int up_cpu_paused(int cpu)
    */
 
   arm_restorestate(tcb->xcp.regs);
-  spin_unlock(&g_cpu_wait[cpu]);
-  spin_unlock(&g_cpu_resumed[cpu]);
 
-  return OK;
+  return 0;
```
Comment (acassis) [inline arch/arm/src/sam34/sam4cm_cpupause.c:222]: Ditto

File: arch/arm64/src/common/arm64_cpupause.c [hunk 11]
```diff
@@ -130,6 +122,38 @@ int up_cpu_paused(int cpu)
 
   arm64_savestate(tcb->xcp.regs);
 
+  return 0;
```
Comment (acassis) [inline arch/arm64/src/common/arm64_cpupause.c:125]: Ditto

File: arch/risc-v/src/common/riscv_cpupause.c [hunk 12]
```diff
@@ -131,6 +123,38 @@ int up_cpu_paused(int cpu)
 
   riscv_savecontext(tcb);
 
+  return 0;
```
Comment (acassis) [inline arch/risc-v/src/common/riscv_cpupause.c:126]: Ditto

File: arch/sim/src/sim/sim_smpsignal.c [hunk 13]
```diff
@@ -174,6 +166,38 @@ int up_cpu_paused(int cpu)
 
   sim_savestate(tcb->xcp.regs);
 
+  return 0;
```
Comment (acassis) [inline arch/sim/src/sim/sim_smpsignal.c:169]: Ditto

File: arch/sim/src/sim/sim_smpsignal.c [hunk 14]
```diff
@@ -214,10 +253,8 @@ int up_cpu_paused(int cpu)
    */
 
   sim_restorestate(tcb->xcp.regs);
-  spin_unlock(&g_cpu_wait[cpu]);
-  spin_unlock(&g_cpu_resumed[cpu]);
 
-  return OK;
+  return 0;
```
Comment (acassis) [inline arch/sim/src/sim/sim_smpsignal.c:257]: Ditto

File: arch/sparc/src/s698pm/s698pm_cpupause.c [hunk 15]
```diff
@@ -131,6 +123,38 @@ int up_cpu_paused(int cpu)
 
   sparc_savestate(tcb->xcp.regs);
 
+  return 0;
```
Comment (acassis) [inline arch/sparc/src/s698pm/s698pm_cpupause.c:126]: Ditto

File: arch/sparc/src/s698pm/s698pm_cpupause.c [hunk 16]
```diff
@@ -163,10 +207,7 @@ int up_cpu_paused(int cpu)
 
   sparc_restorestate(tcb->xcp.regs);
 
-  spin_unlock(&g_cpu_wait[cpu]);
-  spin_unlock(&g_cpu_resumed[cpu]);
-
-  return OK;
+  return 0;
```
Comment (acassis) [inline arch/sparc/src/s698pm/s698pm_cpupause.c:210]: Ditto

File: arch/xtensa/src/common/xtensa_cpupause.c [hunk 17]
```diff
@@ -116,6 +108,38 @@ int up_cpu_paused(int cpu)
 
   xtensa_savestate(tcb->xcp.regs);
 
+  return 0;
```
Comment (acassis) [inline arch/xtensa/src/common/xtensa_cpupause.c:111]: Ditto

File: arch/xtensa/src/common/xtensa_cpupause.c [hunk 18]
```diff
@@ -148,10 +192,7 @@ int up_cpu_paused(int cpu)
 
   xtensa_restorestate(tcb->xcp.regs);
 
-  spin_unlock(&g_cpu_wait[cpu]);
-  spin_unlock(&g_cpu_resumed[cpu]);
-
-  return OK;
+  return 0;
```
Comment (acassis) [inline arch/xtensa/src/common/xtensa_cpupause.c:195]: Ditto

File: arch/arm/src/armv7-a/arm_cpupause.c (hunk 1 above)
Comment (GUIDINGLI) [reply arch/arm/src/armv7-a/arm_cpupause.c]: done

**Review by jerpelea — APPROVED** [review]

**Comment by masayuki2009** [comment]
> @masayuki2009
> 
> Thanks for the feeding back.
> 
> I have updated the PR, please try again.

@GUIDINGLI 
I confirmed that the errors have been fixed.

**Pull request merged**